### PR TITLE
Validate snapshot when snapshot evidence is committed

### PR DESCRIPTION
**Pull request by jumaffre**
Part I of #1539 

Until now, snapshots were written to disk as soon as they were serialised. A new joiner node could then use the snapshot directory and resume from the latest snapshot before joining. However, it was possible for the evidence of the snapshot (committed in a standalone transaction right after the snapshot was generated) to be rolled back (e.g. on election), rendering the snapshot blameless. 

Snapshots stored on disk are now marked as committed (`.committed` extension, similar to ledger chunk files) whenever the evidence for that snapshot has been (globally) committed. New joiners resuming from an existing `snapshots/` directory now skip un-committed snapshot files. 

In other words, if operators and nodes behave non-maliciously, only blame-able snapshots can be used by joiners or for recovery. We still need to add verification of snapshots when new joiners resume from a snapshot: they can only trust the new network once the evidence of the snapshot is validated by a signature transaction (part II of #1539).

### CODE REVIEW

File: src/ds/thread_messaging.h [hunk 1]
```diff
@@ -285,8 +285,9 @@ namespace threading
       return tasks[tid];
     }
 
-    bool run_one(Task& task)
+    bool run_one()
```
Comment (jumaffre) [inline src/ds/thread_messaging.h:288]: I believe it is neater if this function does not take any argument and the task automatically deduced from the current thread ID calling `run_one()`.

File: tests/infra/node.py [hunk 2]
```diff
@@ -255,7 +256,24 @@ def get_ledger(self):
         return self.remote.get_ledger()
 
     def get_snapshots(self):
-        return self.remote.get_snapshots()
+        # Wait for all available snapshot files to be committed before
+        # copying snapshot directory
+        def wait_for_snapshots_to_be_committed(src_dir, list_src_dir_func, timeout=3):
```
Comment (jumaffre) [inline tests/infra/node.py:261]: Because snapshot evidences are committed asynchronously, it is better to wait for all available snapshots to be committed before attempting to resume from one. This function is a little awkward as it has to work for both `Local` and `SSH` remotes which need to pass the `list_src_dir_func` lambda in.

File: tests/infra/node.py [hunk 3]
```diff
@@ -255,7 +256,24 @@ def get_ledger(self):
...
+            while time.time() < end_time:
+                committed = True
+                for f in list_src_dir_func(src_dir):
+                    committed &= f.endswith(".committed")
+                if committed:
+                    break
+                time.sleep(0.1)
+            if not committed:
+                raise RuntimeError(
+                    f"All snapshots were not committed after {timeout}s in {src_dir}"
```
Comment (achamayou) [inline tests/infra/node.py:273]: Should that read "Not all snapshots..."? Printing the set of uncommitted ones may be quite helpful too.

File: tests/infra/remote.py [hunk 4]
```diff
@@ -213,6 +220,10 @@ def get(self, file_name, dst_path, timeout=FILE_TIMEOUT, target_name=None):
                         if os.path.exists(dst_dir):
                             shutil.rmtree(dst_dir)
                         os.makedirs(dst_dir)
+                        if pre_condition_func is not None:
+                            pre_condition_func(
```
Comment (achamayou) [inline tests/infra/remote.py:224]: I would expect a pre_condition_func to cause an operation to abort if it returns false, I guess the idea is to raise here instead?

File: tests/infra/remote.py [hunk 5]
```diff
@@ -189,7 +189,14 @@ def _setup_files(self):
...
         session.close()
 
-    def get(self, file_name, dst_path, timeout=FILE_TIMEOUT, target_name=None):
+    def get(
+        self,
+        file_name,
+        dst_path,
+        timeout=FILE_TIMEOUT,
+        target_name=None,
+        pre_condition_func=None,
```
Comment (achamayou) [inline tests/infra/remote.py:198]:
```suggestion
        pre_condition_func=lambda src_dir, ???: True
```

may be more readable and avoid a branch further down.

**Review by achamayou — APPROVED** [review]

File: src/ds/thread_messaging.h (hunk 1 above)
Comment (ashamis) [reply src/ds/thread_messaging.h:288]: So there is a reason for this, and this change introduces a performance hit on a very hot path

calling get_current_thread_id() is a is a map lookup, this changes running a task from O(1) to O(log(n)).

When this was done we had an issue with thread_local and OE. This should have been resolved after the work Eddy did a couple of months ago. So the correct fix is to create a thread local variable that is this threads task.

**Review by ashamis — CHANGES_REQUESTED** [review]
Incorrect comment

File: src/ds/thread_messaging.h (hunk 1 above)
Comment (eddyashton) [reply src/ds/thread_messaging.h:288]: Afaict, the only place we were calling this `run_one` was from `enclave.h`, where we were also calling exactly the same pattern of functions so incurring the same lookup cost? So this shouldn't be any slower, and looks like a safer API (you can't pass another thread's `Task`).

**Review by ashamis — APPROVED** [review]

File: src/ds/thread_messaging.h (hunk 1 above)
Comment (jumaffre) [reply src/ds/thread_messaging.h:288]: This is followed-up in #1672
